[Driver][ASan] Refactor Clang-Driver "Sanitizer Bitcode" linking.

ampandey-1995 opened the pull request:
ASan bitcode linking is currently available for HIPAMD,OpenMP and OpenCL. Moving sanitizer specific common parts of logic to appropriate API's so as to reduce code redundancy and maintainability.

ampandey-1995 committed:
[Driver][ASan] Refactor Clang-Driver "Sanitizer Bitcode" linking.

ASan bitcode linking is currently available for HIPAMD,OpenMP and
OpenCL. Moving sanitizer specific common parts of logic to appropriate
API's so as to reduce code redundancy and maintainability.

arsenm commented on clang/include/clang/Basic/DiagnosticDriverKinds.td:
```diff
@@ -70,6 +70,10 @@ def err_drv_no_rocm_device_lib : Error<
   "'--rocm-path' or '--rocm-device-lib-path', or pass '-nogpulib' to build "
   "without ROCm device library">;
+def err_drv_no_asan_rt_lib
+    : Error<"AMDGPU address sanitizer runtime library (asanrtl) is not found. "
+            "Please install ROCm device library which supports address "
+            "sanitizer">;
```
```suggestion
    : Error<"address sanitizer runtime library (asanrtl) is not found">;
```

Error messages should start with lowercase, and the suggestion to install anything is strange. Installation of anything is not required. I also don't see why we would have multiple configurations of the libraries, this should just be packaged in 100% of cases with the rest of the libraries

ampandey-1995 replied:
Thanks, @arsenm . We discussed internally and now removing the ```asanrtl bitcode is not found``` diagnostic. 

Hi @arsenm, @yxsamliu, @b-sumner , should I edit the lit test(https://github.com/llvm/llvm-project/blob/bb518655127ad49e527bbc2a57a5fd8a3f9f0495/clang/test/Driver/hip-sanitize-options.hip#L63) with generic diagnostic or remove it also?

arsenm replied:
Should cover it in the test, like all the other libraries

ampandey-1995 replied:
Done.

ampandey-1995 committed:
[OpenMP][ASan] Enable ASan Instrumentation for AMDGPUOpenMPToolChain.

Enable device code ASan instrumentation for openmp offload applications
using option '-fsanitize=address'.

ampandey-1995 committed:
[OpenMP][ASan] Enable ASan Instrumentation for AMDGPUOpenMPToolChain. (#2)

Enable device code ASan instrumentation for openmp offload
applications using option '-fsanitize=address'.

ampandey-1995 committed:
Revert "[OpenMP][ASan] Enable ASan Instrumentation for AMDGPUOpenMPToolChain."

ampandey-1995 committed:
Merge pull request #3 from ampandey-1995/revert-2-omp-asan-driver

Revert "[OpenMP][ASan] Enable ASan Instrumentation for AMDGPUOpenMPToolChain."

ajarmusch commented:
Hi, an LLVM OpenMP Offloading CI/CD detected a new failure during testing on an AMD MI210 and MI300A. The Failed test is [test_requires_atomic_default_mem_order_seq_cst.F90](https://github.com/OpenMP-Validation-and-Verification/OpenMP_VV/blob/master/tests/5.0/requires/test_requires_atomic_default_mem_order_seq_cst.F90).

The error: `error: unknown argument: '-mlink-bitcode-file'`

For full details of the failure see: https://gitlab.e4s.io/uo-public/llvm-openmp-offloading-v2/-/jobs/360327

jhuber6 commented:
> Hi, an LLVM OpenMP Offloading CI/CD detected a new failure during testing on an AMD MI210 and MI300A. The Failed test is [test_requires_atomic_default_mem_order_seq_cst.F90](https://github.com/OpenMP-Validation-and-Verification/OpenMP_VV/blob/master/tests/5.0/requires/test_requires_atomic_default_mem_order_seq_cst.F90).
> 
> The error: `error: unknown argument: '-mlink-bitcode-file'`
> 
> For full details of the failure see: https://gitlab.e4s.io/uo-public/llvm-openmp-offloading-v2/-/jobs/360327

Probably needs to be listed as a FCC1 option.

ampandey-1995 commented:
PR https://github.com/llvm/llvm-project/pull/125322 is up for fixes.

jhuber6 commented on clang/lib/Driver/ToolChains/AMDGPU.cpp:
```diff
@@ -950,6 +950,11 @@ void ROCMToolChain::addClangTargetOptions(
 
+  std::tuple<bool, const SanitizerArgs> GPUSan(
+      DriverArgs.hasFlag(options::OPT_fgpu_sanitize,
+                         options::OPT_fno_gpu_sanitize, true),
```
This cannot be default true, it included `ockl.bc` which breaks OpenMP's printf and includes some other conflicting definitions. Furthermore you disable internalization so they stick around in every single OpenMP program now.

ampandey-1995 replied:
Yes , sorry one misstep is here is that GPUSan object created while not checking that ```-fsanitize=address```  available. This GPUSan should be a pointer and should be initialized when ```-fsanitize=address``` is detected on Args.
